### PR TITLE
i#2509 AArch64 selfmod: Obtain PC from icache instr itself.

### LINKED ISSUE #2509: Segmentation fault in some clients with __clear_cache
On AArch64 some dynamic code operations cause a crash in (or with) clients:
```
#include <stdlib.h>
int main(void) {
    void *page;
    posix_memalign(&page, 4096, 4096);
    __builtin___clear_cache(page, (char*)page + 4096);
    return 0;
}
```
this gets a segmentation fault with simple clients like libbbcount and libopcodes. The debug build says
```
<Application tried to execute from unreadable memory 0x0000000000000000.
```
We know DR on AArch64 traps IC IVAU in order to handle dynamic code, so maybe there's some interaction between that and the instrumentation that these clients are putting in.

### PR DESCRIPTION
**Pull request by egrimley**
When mangling an ISB instruction we need the PC of the following
instruction, which will be the target after flushing fragments.
Previously we used get_app_instr_xl8(next_instr), but the next
instruction might be instrumentation, in which case the translation
fails. So use get_app_instr_xl8(instr) + AARCH64_INSTR_SIZE, instead.

Also, because of the potential side exit, terminate a basic block
on OP_isb, enforce this rule in client_process_bb(), and document
this restriction.

Fixes #2509

Change-Id: Ifcedb70fae7360c791e5857813b4d8d55fdc7471

### CODE REVIEW

**Comment by fhahn** [comment]
run aarch64 tests

**Comment by derekbruening** [comment]
I don't think this will work in all cases.  This assumes unchanged linear code with an observation-only tool, but a client can change app code: a client could remove the instruction after the OP_isb, with the subsequent app instruction no longer at a +4 offset.

Looking at how OP_isb can exit the cache, that is also tricky for tools: they have to avoid keeping unrestored state across it.  I.e., this instruction has to be treated as a barrier, just like syscalls or interrupts are treated by drreg and tool analysis code in general.  Is this behavior of OP_isb documented?  I don't see drreg checking for it.

**Comment by egrimley** [comment]
I think you're right: ISB should be treated like a barrier. Can you point out where drreg checks for syscalls so I can add ISB to the check?

(The way ISB is handled is a consequence of me worrying that the instruction following the ISB might itself have been invalidated, though I've never heard of that happening.)

**Comment by derekbruening** [comment]
> This assumes unchanged linear code with an observation-only tool, but a client can change app code: a client could remove the instruction after the OP_isb, with the subsequent app instruction no longer at a +4 offset.

Actually this part may work out, as a client who wants to remove that subsequent instr would have to do so on the new bb starting there.  It doesn't feel quite right though, I think because it's different from existing mechanisms.  It's closest to an inlined syscall on UNIX, where we have a conditional exit cti, and require clients to treat syscall instrs as potential side exits from bbs.  However, we disable inlined syscalls for -code_api b/c we couldn't easily support clients changing the syscall number when inlined and to avoid side exits in bbs.

Along those lines, thinking through it further, it does not fit with existing syscalls/interrupts/branches for -code_api (i.e., for clients): drreg and other client code is able to assume that such things are bb-final, which is enforced by DR's bb modification rules.  Some of the code would work with them being side exits (esp code acting purely on traces, which also have side exits), but code would have to be added: the trace side exits work b/c drreg & co. do full restores at the end of each bb, which gets propagated to the trace.

So either we have to treat an OP_isb instr like a syscall or x86 interrupt instruction and require that it be bb-final for -code_api, or we have to introduce the possibility of bb side exits for -code_api and add checks (probably also with a new API query named sthg like instr_may_exit()) and add some docs to bt.dox (somewhere in there it talks about linear control flow) to clarify what analysis routines need to do.

**Comment by egrimley** [comment]
rerun aarch64 tests

File: core/arch/interp.c [hunk 1]
```diff
@@ -3920,6 +3920,11 @@ build_bb_ilist(dcontext_t *dcontext, build_bb_t *bb)
             if (!bb_process_interrupt(dcontext, bb))
                 break;
         }
+#ifdef AARCH64
+        /* OP_isb, when mangled, has a potential side exit. */
+        else if (instr_get_opcode(bb->instr) == OP_isb)
```
Comment (derekbruening) [inline core/arch/interp.c]: The requirement that OP_isb is bb-final needs to be a rule for clients, just like the rule that syscalls and interrupts must be bb-final: i.e., please add it to the dr_register_bb_event() docs.

File: api/docs/bt.dox [hunk 2]
```diff
@@ -500,6 +500,8 @@ internal operation requires that we impose the following limitations:
    a basic block, and it must be the final application branch in the block.
  - The exit control-flow of a block ending in a system call cannot be
    changed.
+ - On AArch64, an ISB instruction (OP_isb) must be the last instruction
+   in its block.
```
Comment (derekbruening) [inline api/docs/bt.dox]: The specific docs for dr_register_bb_event() need updating.

Comment (derekbruening) [inline api/docs/bt.dox]: (The docs in its big doxygen comment in instrument_api.h)

File: core/lib/instrument_api.h [hunk 3]
```diff
@@ -221,6 +221,8 @@ DR_API
  * - A system call or interrupt instruction can only be added
  * if it satisfies the above constraints: i.e., if it is the final
  * instruction in the block and the only system call or interrupt.
+ * - Any AArch64 OP_isb instruction must be the last instruction
```
Comment (derekbruening) [inline core/lib/instrument_api.h]: Use `#OP_isb` to get a link.

File: api/docs/bt.dox (hunk 2 above)
Comment (derekbruening) [inline api/docs/bt.dox]: Use `#OP_isb` to get a link.

File: core/lib/instrument_api.h (hunk 3 above)
Comment (egrimley) [inline core/lib/instrument_api.h]: Done.

**Comment by egrimley** [comment]
rerun aarch64 tests

File: core/arch/interp.c [hunk 4]
```diff
@@ -2852,6 +2852,13 @@ client_process_bb(dcontext_t *dcontext, build_bb_t *bb)
         if (!instr_opcode_valid(inst))
             continue;
 
+# ifdef AARCH64
+        if (inst != instrlist_last(bb->ilist)) {
+            CLIENT_ASSERT(instr_get_opcode(inst) != OP_isb,
```
Comment (derekbruening) [inline core/arch/interp.c]: This is too early in the loop body here: it needs to be below the instr_is_meta check below, as a client can insert whatever non-app-instr (i.e., "meta" or "tool") code it wants (we won't mangle it; up to client to get correctness right; might have its own reasons for using it and we shouldn't get in the way).  See how the other checks like instr_is_syscall, client_check_syscall, instr_will_be_exit_cti are down below.

Comment (derekbruening) [inline core/arch/interp.c]: nit: the order seems odd: I would vote for what IMHO is easier to read and matches the existing code: more like "if (instr_get_opcode(inst) == OP_isb) CLIENT_ASSERT(inst == instrlist_last..."

**Comment by egrimley** [comment]
rerun aarch64 tests

**Comment by egrimley** [comment]
The Travis failure looks like a temporary network problem.

**Pull request merged**